client/state/plugins/installed/actions.js

```diff
@@ -0,0 +1,360 @@
+	return site.jetpack ? plugin.id : plugin.slug;
+};
+
+/**
+ * Return a SitePlugin instance used to handle the plugin
+ *
+ * @param {Object} site - site object
+ * @param {String} pluginId - plugin identifier
+ * @return {SitePlugin} SitePlugin instance
+ */
+const getPluginHandler = ( site, pluginId ) => {
+	const siteHandler = wpcom.site( site.ID );
+	const pluginHandler = site.jetpack
+		? siteHandler.plugin( pluginId )
+		: siteHandler.wpcomPlugin( pluginId );
+
+	return pluginHandler;
+};
+
+export function activate( site, plugin ) {
+	return ( dispatch ) => {
+		const pluginId = getPluginId( site, plugin );
+		const defaultAction = {
+			action: ACTIVATE_PLUGIN,
+			siteId: site.ID,
+			pluginId,
+		};
+		dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_ACTIVATE_REQUEST } ) );
+
+		const successCallback = ( data ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_ACTIVATE_REQUEST_SUCCESS, data } ) );
+		};
+
+		const errorCallback = ( error ) => {
+			// This error means it's already active.
+			if ( error && error.error === 'activation_error' ) {
+				successCallback( plugin );
+			}
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_ACTIVATE_REQUEST_FAILURE, error } ) );
+		};
+
+		return getPluginHandler( site, pluginId ).activate().then( successCallback ).catch( errorCallback );
+	};
+}
+
+export function deactivate( site, plugin ) {
+	return ( dispatch ) => {
+		const pluginId = getPluginId( site, plugin );
+		const defaultAction = {
+			action: DEACTIVATE_PLUGIN,
+			siteId: site.ID,
+			pluginId,
+		};
+		dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_DEACTIVATE_REQUEST } ) );
+
+		const successCallback = ( data ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_DEACTIVATE_REQUEST_SUCCESS, data } ) );
+		};
+
+		const errorCallback = ( error ) => {
+			// This error means it's already inactive.
+			if ( error && error.error === 'deactivation_error' ) {
+				successCallback( plugin );
+			}
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_DEACTIVATE_REQUEST_FAILURE, error } ) );
+		};
+
+		return getPluginHandler( site, pluginId ).deactivate().then( successCallback ).catch( errorCallback );
+	};
+}
+
+export function update( site, plugin ) {
+	return ( dispatch ) => {
+		if ( ! plugin.update ) {
+			return Promise.reject( 'Error: Plugin already up-to-date.' );
+		}
+
+		const pluginId = getPluginId( site, plugin );
+		const defaultAction = {
+			action: UPDATE_PLUGIN,
+			siteId: site.ID,
+			pluginId,
+		};
+		dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_UPDATE_REQUEST } ) );
+
+		const successCallback = ( data ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_UPDATE_REQUEST_SUCCESS, data } ) );
+		};
+
+		const errorCallback = ( error ) => {
+			dispatch( Object.assign( {}, defaultAction, { type: PLUGIN_UPDATE_REQUEST_FAILURE, error } ) );
+		};
+
+		if ( ! site.canUpdateFiles ) {
```

Comment:
To earlier point on passing `site` object, to workaround it here, we'd typically not have validation in an action creator but instead at the point where the action creator would be dispatched. The boilerplate is usually avoided by consolidating it within a query component, though it's a little less obvious how we'd handle that for update actions like this.